File: year_2023/day_02/day02.py
import re
from dataclasses import dataclass

from tools import loader, parsers


@dataclass
class Bag:
    ID: int
    red: int = 0
    green: int = 0
    blue: int = 0

    def is_valid(self) -> bool:
        return self.red <= 12 and self.green <= 13 and self.blue <= 14

    def add_cube(self, colour: str, amount: int) -> None:
        setattr(self, colour, max(getattr(self, colour), amount))

    def power(self) -> int:
        return self.red * self.green * self.blue


def cubes(data: list[str]) -> tuple[int, int]:
    """
    >>> print(cubes(parsers.lines('test.txt')))
    (8, 2286)"""
    bags = []
    for line in data:
        bag = Bag(ID=int(re.findall(r'\d+', line)[0]))
        for cube in re.finditer(r'(\d+) (\w+)', line):
            bag.add_cube(cube.group(2), int(cube.group(1)))
        bags.append(bag)
    part1 = sum(bag.ID for bag in bags if bag.is_valid())
    part2 = sum(bag.power() for bag in bags)
    return part1, part2


print(cubes(parsers.lines(loader.get())))  # 2369, 66363

File: year_2023/day_02/day02.go
package main

import (
	"aoc/tools"
	"fmt"
	"regexp"
	"strconv"
)

func solve(input string) (int, int) {
	part1, part2 := 0, 0
	re := regexp.MustCompile(`(\d+) (\w+)`)
	nums := regexp.MustCompile(`\d+`)

	for _, line := range tools.ReadLines(input) {
		b := map[string]int{"red": 0, "green": 0, "blue": 0}
		cubes := re.FindAllStringSubmatch(line, -1)

		for _, cube := range cubes {
			value, _ := strconv.Atoi(cube[1])
			b[cube[2]] = max(value, b[cube[2]])
		}

		if b["red"] <= 12 && b["green"] <= 13 && b["blue"] <= 14 {
			index, _ := strconv.Atoi(nums.FindString(line))
			part1 += index
		}

		part2 += b["red"] * b["green"] * b["blue"]
	}

	return part1, part2
}

func main() {
	fmt.Println(solve("test.txt"))
	fmt.Println(solve(tools.GetData(2023, 2)))
}
